javascript/0080-remove-duplicates-from-sorted-array-ii.js

```diff
@@ -1,22 +1,61 @@
 /**
- * Linear 
- * Time O(N) | Space O(1)
- * https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
- * @param {number[]} nums
- * @return {number}
- */
+* https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
+* 
+* Time O(n) | Space O(1)
+* @param {number[]} nums
+* @return {number}
+*/
 var removeDuplicates = function(nums) {
-    if (nums.length < 3) {
-        return nums.length;
+  let current = nums[0];
+  let sameElCount = 0;
+
+  for(let i = 0; i < nums.length; i++) {
+    if(current === nums[i]) {
+        sameElCount++;
+    } 
+    if(current !== nums[i]) {
+        current = nums[i];
+        sameElCount = 1;
     }
-    let l = 1;
-    let r = 2;
-    while (r < nums.length) {
-        if (nums[l] !== nums[r] || (nums[l] === nums[r] && nums[l-1] !== nums[r])) {
-            l += 1;
-            nums[l] = nums[r];
-        }
-        r += 1;
+   if(sameElCount > 2) {
+       nums.splice(i,1);
+       i--;
     }
-    return l + 1;
+  }
 };
+
+
+/**
+* Two pointer
+* Time O(n^2) | Space O(1)
+* @param {number[]} nums
+* @return {number}
+*/
+var removeDuplicates2 = function(nums) {
+
+  let current = nums[0];
+  let sameElCount = 0;
+
+
+  for(let i = 0; i < nums.length; i++) {
+    if(current === nums[i]) {
+        sameElCount++;
+    } 
+    if(current !== nums[i]) {
+        current = nums[i];
+        sameElCount = 1;
+    }
+    if(sameElCount > 2) {
+        let left = i;
+        let right = i+1;
+        let count = 1;
+        while(nums[left] === nums[right]) {
+            count++;
+            right++;
+        }
+        nums.splice(left, count);
```

Comment:
You can use 2-pointer to avoid using splice

```js

/**
 * 2-pointer
 * Time O(N) | Space O(1)
 * https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = (nums) => {
    const isEdgeCase = (nums.length < 2)
    if (isEdgeCase) return nums.length;

    let [ left, right ] = [ 2, 2 ];

    while (right < nums.length) {/* Time O(N) */
        const isEqual = (nums[(left - 2)] === nums[right]);
        if (!isEqual) {
            nums[left] = nums[right];
            left += 1;
        }

        right += 1;
    }

    return left;
};

```

Reply:
done